File: app/src/main/java/com/example/movieapi/api/NetworkService.java
package com.example.movieapi.api;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class NetworkService {

    private static NetworkService instance;
    private String BASE_URL = "https://api.themoviedb.org";
    private Retrofit retrofit;

    private NetworkService() {
        retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
    }

    public static NetworkService getInstance() {
        if (instance == null) {
            instance = new NetworkService();
        }
        return instance;
    }

    public APIService getAPIService() {
        return retrofit.create(APIService.class);
    }

}

File: app/src/main/java/com/example/movieapi/api/NetworkService.kt
package com.example.movieapi.api

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class NetworkService private constructor() {
    private val baseUrl = "https://api.themoviedb.org"

    private val retrofit: Retrofit

    val aPIService: APIService
        get() = retrofit.create(APIService::class.java)

    companion object {
        var instance: NetworkService? = null
            get() {
                if (field == null) {
                    field = NetworkService()
                }
                return field
            }
            private set
    }

    init {
        retrofit = Retrofit.Builder()
                .baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
    }
}